python/models/hierarchical_filtering_proxy_model.py

```diff
@@ -329,104 +333,122 @@ def setSourceModel(self, model):
         if cached_value is not None:
-            # we already computed this so just return the result
-            return cached_value
-
-        # need to recursively iterate over children looking for one that is accepted:
-        child_accepted = False
-        for ci in range(model.rowCount(idx)):
-            child_idx = idx.child(ci, 0)
-
-            # Check if child item is in cache:
-            child_accepted = self._accepted_cache.get(child_idx)
-            if child_accepted is None:
-                # It's not in the cache so lets see if it's accepted and add to the cache:
-                child_accepted = self._is_row_accepted(child_idx.row(), idx, parent_accepted)
-                self._accepted_cache.add(child_idx, child_accepted)
-
-            if not child_accepted and model.hasChildren(child_idx):
-                # This child was not accepted.
-                # Recurse down and check if any grand children are accepted.
-                child_accepted = self._is_child_accepted_r(child_idx, False)
-
-            if child_accepted:
-                # This child node or one of its descendants indicated that they
-                # were accepted so exit early.
-                break
-
-        # Cache if one of the descendants was accepted or not.
-        self._child_accepted_cache.add(idx, child_accepted)
-        return child_accepted
+            return not cached_value
+        # We don't have a cache value check if this item can be accepted
+        accepted = self._is_row_accepted(from_index.row(), from_index.parent(), False)
+        self._rejected_cache.add(from_index, not accepted)
+        if accepted:
+            # Great, no need to check further.
+            return True
+        # We need to check if we should be accepted because one of our
+        # descendant is accepted.
+        # We keep a list of indexes to check, add new indexes at the end and
+        # unqueue them from the left (first in first out).
+        descendant_indexes = deque()
+        for row_i in range(from_index.model().rowCount(from_index)):
+            descendant_indexes.append(from_index.child(row_i, 0))
+        while descendant_indexes:
+            child_index = descendant_indexes.popleft()
+            rejected = self._rejected_cache.get(child_index)
+            if cached_value is None:
+                # Compute the value
+                accepted = self._is_row_accepted(
+                    child_index.row(),
+                    child_index.parent(),
+                    False
+                )
+                rejected = not accepted
+                self._rejected_cache.add(child_index, rejected)
+            if rejected:
+                # We need to check children, if any, add them in the
+                # descendant list.
+                for row_i in range(child_index.model().rowCount(child_index)):
+                    descendant_indexes.append(child_index.child(row_i, 0))
```

Comment:
why not make the method recursive? i personally find this pretty hard to read. Is it an optimization of sorts?

Reply:
it is not recursive because it does not need to be: we don't need to get back results from children recursions. The cache is updated bottom to top around line 409, and the update can go higher in hierarchy than the starting point. Usually iterative approaches are more efficient than recursive approaches. When comparing the previous approach with this one, I didn't spot any noticeable differences for a Sequence/Shot/Step/Task tree, ~ 5000 Tasks, so I guess it could be turned into a recursive approach if you feel strongly about it?

Reply:
not feeling strongly! Whatever you feel is best!

Reply:
Steph, did you test with something a bigger than 5000 tasks and see how your code scales with bigger payloads? I remember originally the time to build the tree used to be exponential in wf2.